New Resource: aws_kms_grant

jammerful opened the pull request:
Hey,

Basically I think this is close to being ready to merge, but I see two flaws with it:

* Dealing with the `InvalidArnException` upon creation of a grant that depends on a newly created role. I have a retry in my code, but it will just retry in the case that the arn is actually invalid. I don't see a way to distinguish between eventual consistency and a truly invalid arn.
  - I think this maybe okay behavior.
* In a similar vein, if you refresh the state quickly after creating a role and grant (as is the case with the acceptance tests currently). The read function will sometimes get an internal identifier (AROAXXXXX instad of arn:aws:XXX) to the grantee principal instead of an ARN, which would if you naively take the value be written to the state file and cause a new resource to be created since grants have no way of being updated. 
  - What I did, which I saw done in another resource file, is if the read doesn't return valid grant or retiree principals then don't update the state file for those attributes and log a warning. This will only happen in two cases ~1 minute after iam principal creation (aws acknoleged eventual consistency problem) and if one of the principals has been deleted (in cases where it is managed outside of terraform).

If merged this resolves #641

I have run the acceptance tests in their current form, and they pass:

```
$ make testacc TESTARGS='-run=TestAWSKmsGrant'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAWSKmsGrant -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAWSKmsGrant_Basic
--- PASS: TestAWSKmsGrant_Basic (39.08s)
=== RUN   TestAWSKmsGrant_withConstraints
--- PASS: TestAWSKmsGrant_withConstraints (30.75s)
=== RUN   TestAWSKmsGrant_withRetiringPrincipal
--- PASS: TestAWSKmsGrant_withRetiringPrincipal (39.39s)
=== RUN   TestAWSKmsGrant_bare
--- PASS: TestAWSKmsGrant_bare (38.95s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	148.205s
```

I have also run the unit tests and they pass:

```
$ make test
==> Checking that code complies with gofmt requirements...
go test ./... -timeout=30s -parallel=4
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1.765s
```

jammerful commented:
Resolves #641

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,414 @@
+				// an InvalidArnException to be thrown. TODO: Possibly change the aws_iam_role code?
+				if awsErr.Code() == "DependencyTimeoutException" ||
+					awsErr.Code() == "InternalException" ||
+					awsErr.Code() == "InvalidArnException" {
```
This is the first issue I mentioned.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,414 @@
+		// Confirmed with AWS that this is expected behavior.
+		// TODO: Handle upstream or here?
+		if grant != nil {
+			if !strings.HasPrefix(*grant.GranteePrincipal, "arn:aws") {
```
This is the second issue I mentioned.

jammerful commented:
@radeksimko @bflad Can you please take a look at this soon?

kikitux commented:
Sorry for the delay, finding someone for a review

paddycarver commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+		grant, err := findKmsGrantById(conn, keyId, grantId, nil)
+		if err != nil {
+			if _, ok := err.(KmsGrantMissingError); ok {
+				if grant == nil {
```
In what condition would the grant be non-nil and the error be non-nil?

jammerful replied:
There isn't one, I'll remove the innermost if statement and just return nil (see findKmsGrantById).

paddycarver commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+		m["encryption_context_subset"] = pointersMapToStringList(constraint.EncryptionContextSubset)
+	}
+
+	return []interface{}{m}
```
Can we only have one constraint? Why does the expand version loop over it then?

jammerful replied:
Only one of encryption_context_subset or encrytpion_context_equals can be set, but that itself can be a map. Will have to refactor this as part of the move to sets.

paddycarver commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+
+// Custom error, so we don't have to rely on
+// the content of an error message
+type KmsGrantMissingError struct {
```
This could just be a string. `type KmsGrantMissingError string`.

jammerful replied:
Will fix.

paddycarver commented on aws/resource_aws_kms_grant_test.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+resource "aws_kms_grant" "%s" {
+	name = "%s"
```
does this need to be unique?

paddycarver replied:
Awesome, thanks! I just wanted to be sure we wouldn't run into an error if the resource got orphaned, which we usually try to avoid using randomized names. Sounds like we don't need a randomized name, here, though.

paddycarver commented on aws/validators.go:
```diff
@@ -1395,6 +1395,43 @@ func validateAwsKmsName(v interface{}, k string) (ws []string, es []error) {
+
+	if _, ok := validTypes[value]; !ok {
+		es = append(es, fmt.Errorf(
+			"%s must be one of the following: [ Decrypt, Encrypt, GenerateDataKey, GenerateDataKeyWithoutPlaintext, ReEncryptFrom, ReEncryptTo, CreateGrant, RetireGrant, DescribeKey ]", k))
```
Could we programmatically generate this list, to prevent drift?

jammerful replied:
Will pull from the above map above thanks. By the way this kind of duplication is pretty common in this file, is probably worth cleaning up.

kikitux replied:
lets follow the suggestion in this PR, your suggestion is welcome but lets focus the conversation on the PR it self 

alvaro

jammerful replied:
No problem, was just bringing a small possible issue to your attention.

paddycarver commented on aws/validators.go:
```diff
@@ -1395,6 +1395,43 @@ func validateAwsKmsName(v interface{}, k string) (ws []string, es []error) {
+func validateAwsKmsGrantOperation(v interface{}, k string) (ws []string, es []error) {
+	value := v.(string)
+
+	validTypes := map[string]bool{
```
This could be a `map[string]struct{}` with each value set to `struct{}{}`, just to denote that the value of the map has no meaning.

jammerful replied:
Why is this better than map[string]bool? Just to denote the value of the map has no meaning?

paddycarver commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+	}
+	if grant != nil {
+		// The grant sometimes contains principals that identified by their unique id: "AROAJYCVIVUZIMTXXXXX"
+		// instead of "arn:aws:...", in this case don't update the state file
```
Oh, I see, you acknowledged this in the PR body. In this case, I think it'd be better to retry until we get an ID in the `arn:aws:` format (or a max threshold is hit) than to not set state.

jammerful replied:
We could store the internal ID. 
But think of this case: the only thing is that typically Create returns the Read function, which will in cases of new IAM roles cause internal IDs to be stored. Then when lets say some just applies terraform again, the change from internal ID to ARN will cause new resources to be created.

Let me ensure that there isn't something clever I'm missing, I'll ask a few people I know that have dealt with these eventual consistency issues. Would it be possible for you to ask a few co-workers as well?

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+				ValidateFunc: validateArn,
+			},
+			"operations": {
+				Type: schema.TypeList,
```
The answer to all of these is no. Will make a set, I think the real benefit here is the set avoids duplication which could cause errors upon creation.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+				ForceNew: true,
+			},
+			"constraints": {
+				Type:     schema.TypeList,
```
Will make a set.

jammerful replied:
I'm almost done with my revision, so I'll update the PR tomorrow. Sorry for the delay, the hash function for constraints now that it is a set is causing me a little heartache.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+							Type:          schema.TypeMap,
+							Optional:      true,
+							Elem:          schema.TypeString,
+							ConflictsWith: []string{"constraints.0.encryption_context_subset"},
```
It worked it my testing, I'm trying not allow both encryption_context_subset and encryption_context_equals both set, since the AWS KMS Grant API does not allow that. I'll have to change how that looks with the move to sets. Make sense?

paddycarver replied:
Yeah, I think I get what you were trying to do here, the syntax is just a bit :/ It may be worthwhile to just implement that logic manually to throw an error in the Create function.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+				ValidateFunc: validateArn,
+			},
+			"grant_creation_tokens": {
+				Type:     schema.TypeList,
```
Will move to a set, as the tokens should be unique. I don't see this a generating a lot of value, but I'll move it to a set.

paddycarver replied:
The main value is that if the API ever changes the order, or the user changes the order, the grant doesn't need to be recreated for no real purpose.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+			} else {
+				log.Printf(
+					"[WARN] Unable to update retiring principal state %s for grant id %s for key id %s",
+					*grant.RetiringPrincipal, grantId, keyId)
```
See the above comment.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+			}
+		}
+
+		d.Set("operations", grant.Operations)
```
Will fix, thanks.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+			d.Set("name", grant.Name)
+		}
+		if grant.Constraints != nil {
+			d.Set("constraints", flattenKmsGrantConstraints(grant.Constraints))
```
Will fix, thanks.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+		}
+
+		d.Set("operations", grant.Operations)
+		if *grant.Name != "" {
```
No, the grant name is optional, however if you don't set it AWS will set it to "". However, if I was to  load that into the state it would cause the state to mismatch the specified state which would cause resource re-creation. So if the grant name is empty just ignore it.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+		if grant.Constraints != nil {
+			d.Set("constraints", flattenKmsGrantConstraints(grant.Constraints))
+		}
+	} else {
```
Will fix, thanks.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+
+	if err != nil {
+		if grant != nil {
+			return true, err
```
Will fix, thanks.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,418 @@
+}
+
+func getKmsGrantById(grants []*kms.GrantListEntry, grantIdentifier string) *kms.GrantListEntry {
+	for idx := range grants {
```
Will fix, thanks.

jammerful commented:
@paddycarver Thanks for the review. I left some comments, most of your concerns will be fixed. Except for the one about not loading in IAM principals into terraform state if it doesn't resolve to an ARN--I left some reasons on why retrying isn't optimal (maybe if we do something like if d.isNewResource but even then you really want that to be on the IAM principal not if the grant is new).

I did have one other concern, about the scalability of this approach. I was stress testing my code, and it can handle about ~200 grants managed in one resource file, if it is more the resource refreshing takes too long. The reason for the limit, is that for each resource/grant we are calling list grants, which will get throttled by AWS. Would it make sense to have a global list of grants? Have you seen other resources implement this (resources that don't have a Describe* call)? I could see this getting messy.

paddycarver commented:
Thanks for the work on this and the careful explanations! For the scale part... The only solution I can think of at the moment is to introduce user-specified timeouts, like this: https://github.com/terraform-providers/terraform-provider-aws/blob/177a19851fbbf8ec6a54f6733312ae205d7327d3/aws/resource_aws_ami_from_instance.go#L30-L34

But I'm toying with the idea of just waiting to see if people hit that timeout case, and then adding the custom timeouts if that turns into an issue people have. I don't know how common it is, unfortunately, to create 200 grants.

There were murmurs and talks of eventually adding some sort of request batching capabilities to Terraform, but I don't think that's something people were talking about for the near future. Though that would ideally be the "real" solution to this, reusing a single List call across resources.

Unfortunately, I think trying to build batching in using a single List call at the moment would be buggy and error-prone, so I'd advise against it unless we discover we absolutely must implement it.

jammerful commented:
I may add user specific timeouts, is that fine? I think I'll have a use for them myself.

jammerful commented:
@paddycarver Addressed your comments. Please take a look.

Regarding sometimes not setting the roles in state, the only thing I could think of is to always store the role id in the state file. However, I this won't work if the role is from another account since the caller won't have GetRole permissions on that role from another account.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,479 @@
+
+// The hash needs to encapsulate what type of constraint it is
+// as well as the keys and values of the constraint.
+func resourceKmsGrantConstraintsHash(v interface{}) int {
```
I need to fix this up. As of now, I know I at least need to sort the maps so that the constraint gets consistent hashes. I couldn't find any common utilities for this.

jammerful replied:
Fixed.

jammerful commented:
@paddycarver I think all of the initial comments have been resolved. Please take another look. 

I still don't have a sense if not setting the roles in state is a non-starter/blocker. Keep in mind there is no way to modify a grant, so I wouldn't be extremely concerned about drift between the state file and what is in AWS.

jammerful commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+		input.Name = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("constraints"); ok {
+		if !kmsGrantConstraintsIsValid(v.(*schema.Set)) {
```
Not sure how much value this is delivering, I really liked when we used conflicts with in the schema because it validates this before going out to AWS. Right now this is caught right before AWS would let us know of the conflict, after other resources have already been created.

paddycarver replied:
I'd love this to fail at plan time, too, but in the interest of moving this forward, I wanted to keep our scope here simple and not introduce complicated things like CustomDiff or ConflictsWith on a Set, which _I've_ never used on a Set before, and don't know the intricacies of how it works. @bflad or @paultyng may have more insight on that.

It's almost certainly possible to have this fail at plan time. I wouldn't consider making that happen a blocker, but if you'd like it to be in the first pass, I'm happy to have it included.

jammerful replied:
Will address in a follow up PR, as I mentioned in another comment. Might make the most sense to move this back to a list? The set adds a lot of additional code, as well as not adding much value I think given that there can only be one element.

jammerful commented:
@paddycarver @kikitux If you can't look at this soon could I please get an indication of whether you are inclined to merge this, or are there major issues preventing this from being merged?

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+				Set:  schema.HashString,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validateAwsKmsGrantOperation,
```
We can avoid a custom validation function and associated unit test by using Terraform core's `validation.StringInSlice()` 😄  (it should also use SDK provided constants)

```go
ValidateFunc: validation.StringInSlice([]string{
  kms.GrantOperationCreateGrant,
  kms.GrantOperationDecrypt,
  kms.GrantOperationDescribeKey,
  kms.GrantOperationEncrypt,
  kms.GrantOperationGenerateDataKey,
  kms.GrantOperationGenerateDataKeyWithoutPlaintext,
  kms.GrantOperationReEncryptFrom,
  kms.GrantOperationReEncryptTo,
  kms.GrantOperationRetireGrant,
}, false),
```

jammerful replied:
Done.

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+		out, err = conn.CreateGrant(&input)
+
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok {
```
You can avoid dealing with `awserr` directly by using the `isAWSErr()` helper function we have. 😄  (similarly the SDK usually provides constants for the error codes)

```go
if isAWSErr(err, kms.ErrCodeDependencyTimeoutException, "") {
```

jammerful replied:
Done.

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+	if grant == nil {
+		log.Printf("[WARN] %s KMS grant id not found for key id %s, removing from state file", grantId, keyId)
+		d.SetId("")
+	}
```
This should `return nil` so the `if grant != nil {` below (and its nesting) is unnecessary.

jammerful replied:
Done.

copumpkin commented:
> Unfortunately, I think trying to build batching in using a single List call at the moment would be buggy and error-prone, so I'd advise against it unless we discover we absolutely must implement it.

@paddycarver if I understand correctly, without batching we get basically quadratic (actually `O(m * n)` where `m` is number of TF grant resources and `n` is number of grants in the account, which after the first time is always `>= m`). My concern here is that unlike other resources with similar issues (I think KMS aliases), grants are often quite numerous. One of my keys today has several hundred grants on it, the default limit per CMK is 2500 (and I've had it raised by AWS support in the past), and I'd like to manage many of those grants with KMS. That seems like it'll become untenable pretty quickly, unless you have a small handful of grants. But with a small number of grants, you might as well just encode it in the key policy.

I get that Terraform doesn't have a proper generic solution for this problem yet, but it doesn't seem awful to maintain a bit of state across grant resource calls so they don't each have to list grants from scratch.

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+	return nil
+}
+
+// Retiring grants requires special permissions (i.e. the
```
Seems like this behavior should be noted in the documentation. Better yet, it would probably be best to provide an attribute flag to control the behavior (e.g. `retire_on_delete`) that defaults to `true` so we can always return an error if an expected deletion (retire or just revoke) is not possible.

jammerful replied:
Done. However, I inverted the default to false. Functionally retiring and revoking are the same, they don't show up in the ListGrants call, however retiring requires special permissions. It seemed to me that defaulting to that would be noisy and cause confusion.

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+		return err
+	}
+
+	d.SetId("")
```
Nitpick: `d.SetId("")` is no longer required in delete functions

jammerful replied:
Done. Didn't realize that.

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+	_, err := conn.RevokeGrant(&input)
+
+	if err != nil {
+		return err
```
We should allow non-existent grants to skip returning an error so its not required to do a `terraform state rm`:

```go
if isAWSErr(err, kms.ErrCodeNotFoundException, "") {
  return nil
}
```

jammerful replied:
Great catch. Done.

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+
+// Custom error, so we don't have to rely on
+// the content of an error message
+type KmsGrantMissingError string
```
Instead of creating our own error class, can we instead utilize the upstream SDK's `awserr.Error` and set to a code like `kms.ErrCodeNotFoundException`?

jammerful replied:
`kms.ErrCodeNotFoundException` is a string, and I wanted a type or code to use so I don't have to key off the error message. Might be missing something here?

bflad commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,502 @@
+	}
+
+	log.Printf("[DEBUG] Created new KMS Grant: %s", *out.GrantId)
+	d.SetId(*out.GrantId)
```
Given that grant IDs are really under a key ID, we might want to consider making the resource ID in Terraform contain both, e.g. `KeyID:GrantID`. This will allow us to import them.

Saving as part of the resource identifier:

```go
// Using keyID := d.Get("key_id").(string) above input
d.SetId(fmt.Sprintf("%s:%s", keyID, aws.StringValue(out.GrantId)))
```

Parsing anytime necessary:

```
keyID, grantID, err := decodeKmsGrantId(d.Id())
if err != nil {
  return err
}
// ...
func decodeKmsGrantId(id string) (string, string, error) {
	parts := strings.Split(id, ":")
	if len(parts) != 2 {
		return "", "", fmt.Errorf("Unexpected format of ID (%q), expected KeyID:GrantID", id)
	}
	return parts[0], parts[1], nil
}
```

jammerful replied:
Done.

paultyng commented:
@copumpkin I think the issue is what if there is a grant that depends on a grant, etc, things like that? If you needed the token to create another grant for example then that cache starts to work against you, which is why it hasn't been used in other places.

I think since the grants are listed by `key` you could just create an `aws_kms_grants` resource that worked in bulk potentially where the `ID` was the `key`?

copumpkin commented:
@paultyng that's an idea that occurred to us, but is there precedent for it in Terraform?

Also, is there a good place (like a TF contributors Slack) to chat live about this sort of thing? Might help sort through these concerns more quickly. There's a few options that have occurred to us but it's hard to factor in what might work vs. trying to guess what HashiCorp might be willing to merge.

paddycarver commented on aws/resource_aws_kms_grant.go:
```diff
@@ -0,0 +1,540 @@
+	return nil, NewKmsGrantMissingError(fmt.Sprintf("[DEBUG] Grant %s not found for key id: %s", grantId, keyId))
+}
+
+// Can't have both constraint options set:
```
I'll admit I'm confused by this part. You're only allowed to have 1 constraint, right? If you're only allowed 1 constraint, is the only reason constraints is a TypeSet for the block syntax? If that's the case, we _could_ use `MaxItems: 1` on `constraints` to make sure no more than 1 block is specified. Doesn't obviate the need for this function, but can simplify it a little.

**Note:** This is not a blocker for merge.

jammerful replied:
Setting `MaxItems` won't work given the way the constraint set is being hashed, it ends up being one item with two different sub-attributes. Maybe I should have left it as a list, mostly had moved it on your suggestion.

paddycarver commented on website/docs/r/kms_grant.html.markdown:
```diff
@@ -0,0 +1,75 @@
+
+* `name` - (Optional, Forces new resources) A friendly name for identifying the grant.
+* `key_id` - (Required, Forces new resources) The unique identifier for the customer master key (CMK) that the grant applies to.
+* `grantee_principal` - (Required, Forces new resources) The principal that is given permission to perform the operations that the grant permits.
```
Could we document here the eventual consistency problem and workaround we have in place, with the ARN weirdness?

jammerful replied:
Done.

jammerful commented:
@paddycarver Thank you for all the help and comments. I updated the documentation, and replied to your comments. Should be ready to merge. Let me know if you want me to squash the commits.

paddycarver commented:
Looks great! Thanks for contributing to Terraform's AWS provider!

copumpkin commented:
Sweet, thanks! I also have an idea for how to get rid of the `O(m * n)` behavior that I'll propose in a separate issue and tag you on, if you don't mind.

paddycarver commented:
Feel free to tag me, and I'll do my best to get to it, but to set expectations appropriately: I am responsible for other providers, so it may take me a bit to get to it; this week is already looking unlikely. There's no harm in tagging me on it, but your best bet for getting eyes on it is to open it up to the maximum number of maintainers. :)

bflad commented:
This has been released in [version 1.12.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1120-march-23-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
